### PR TITLE
feat: unify balancer fetchers

### PR DESCRIPTION
**Pull request by cehali**

### CODE REVIEW

File: packages/oracle-node/src/fetchers/balancer/all-balancers-fetchers.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import { balancerConfigs } from "./balancer-fetcher-configs";
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/all-balancers-fetchers.ts:1]: Why changing file name to plular 'balancer`s`'

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts [hunk 2]
```diff
@@ -1,92 +1,113 @@
...
+  sender: AddressZero,
+  recipient: AddressZero,
+  fromInternalBalance: false,
+  toInternalBalance: false,
 };
 
-export interface BalancerResponse {
-  pool: PoolWithMethods;
-  pairedTokenPrice: number;
+type BalancerPoolsConfigs = Record<
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:19]: I think it would be more readable to create a type `BalancerPoolConfig` as this big object (the second argument to `Record`) and then define `BalancerPoolsConfigs` as
`[token: string]: BalancerPoolConfig`. This way you would have names for things 
 that hide behind `Record` arguments, especially for the current version of this type it is not obvious what this first argument `string` meaning is.

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts [hunk 3]
```diff
@@ -1,92 +1,113 @@
...
+  string,
+  {
+    tokenIn: string;
+    tokenOut: string;
+    swapAmount: BigNumber;
+    swapAmountForSwaps: BigNumber;
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:25]: What is the purpose of this field?

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts [hunk 4]
```diff
@@ -1,92 +1,113 @@
...
 
-  protected getPoolIdForAssetId(assetId: string) {
-    const poolId = this.config.poolsConfigs[assetId].poolId;
-    if (!poolId) {
-      throw new Error(
-        `Missing balancer pair for ${assetId}, check balancer pair config`
-      );
-    }
-    return poolId;
+  override calculateLiquidity(
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:105]: Default implementation already terminates the application with the appropriate error. We don't need to override it I think.

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        "0x6b175474e89094c44da98b954eedeac495271d0f",
+      ],
+      marketSp: "1.019442252013350645857275711595224196226780743325017064",
+      tokenInForSwaps: "0x40d16fc0246ad3160ccc09b8d0d3a2cd28ae6c2f",
+      tokenOutFromSwaps: "0x6b175474e89094c44da98b954eedeac495271d0f",
+    },
+    OHM: {
+      tokenIn: "0x64aa3364f17a4d01c6f1751fd97c2bd3d7e7f1d5", // OHM
+      tokenOut: "0x6b175474e89094c44da98b954eedeac495271d0f", // DAI
+      priceDivider: 10e8,
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:164]: Where the number `10e8` comes from? How is it related to the difference in decimals of tokens involved?

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+          userData: "0x",
+        },
+      ],
+      tokenAddresses: [
+        "0x40d16fc0246ad3160ccc09b8d0d3a2cd28ae6c2f",
+        "0xc443c15033fcb6cf72cc24f1bda0db070ddd9786",
+        "0xfa24a90a3f2bbe5feea92b95cd0d14ce709649f9",
+        "0x6b175474e89094c44da98b954eedeac495271d0f",
+      ],
+      marketSp: "1.019442252013350645857275711595224196226780743325017064",
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:157]: What is this property about?

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts [hunk 7]
```diff
@@ -0,0 +1,209 @@
...
+        "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
+      ],
+
+      tokenInForSwaps: "0xf951e335afb289353dc249e82926178eac7ded78",
+      tokenOutFromSwaps: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
+    },
+    OHM: {
+      tokenIn: "0x64aa3364f17a4d01c6f1751fd97c2bd3d7e7f1d5", // OHM
+      tokenOut: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2", // WETH
+      priceDivider: 10e8,
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:49]: How is this related to the difference in tokens' decimals?

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts [hunk 8]
```diff
@@ -0,0 +1,209 @@
...
+const BIGGER_AMOUNT = utils.parseUnits("1", BIGGER_DECIMALS);
+const SMALLER_AMOUNT = utils.parseUnits("1", SMALLER_DECIMALS);
+
+export const balancerConfigs = {
+  WETH: {
+    SWETH: {
+      tokenIn: "0xf951e335afb289353dc249e82926178eac7ded78", // swETH
+      tokenOut: "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2", // WETH
+      tokenToFetch: "ETH",
+      swapAmount: DEFAULT_AMOUNT,
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:17]: How do we figure out what amount to swap for each pair?

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts (hunk 3 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:25]: Those are parameters required to simulate swaps through balancer pools

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts [hunk 9]
```diff
@@ -1,92 +1,113 @@
...
+type BalancerPoolsConfigs = Record<
+  string,
+  {
+    tokenIn: string;
+    tokenOut: string;
+    swapAmount: BigNumber;
+    swapAmountForSwaps: BigNumber;
+    swaps: SwapConfig[];
+    tokenAddresses: string[];
+    tokenInForSwaps: string;
```
[the remark this answers is not in the document]
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:28]: Those parameters generated through balancer SDK let us simulate swaps on balancers pools

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts (hunk 8 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:17]: You put it in the script for generating those parameters: https://github.com/redstone-finance/redstone-oracles-monorepo/blob/main/packages/oracle-node/tools/balancer/generate-swaps-config.ts. Amount values were selected with Kuba as giving the best results

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts (hunk 6 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:157]: It is parameter generated by script: https://github.com/redstone-finance/redstone-oracles-monorepo/blob/main/packages/oracle-node/tools/balancer/generate-swaps-config.ts

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts (hunk 7 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:49]: It isn't, it would be 10e9 because OHM has 9 decimals and WETH 18 decimals. I have to add it because results from the simulated swap through the contract were correct but multiplied by 10e8

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts (hunk 5 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:164]: I have to add it because results from the simulated swap through the contract were correct but multiplied by 10e8

File: packages/oracle-node/src/fetchers/balancer/all-balancers-fetchers.ts (hunk 1 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/all-balancers-fetchers.ts:1]: probably misclick

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts (hunk 3 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts:25]: I see what you mean, yes you are right. Previously we passed the whole config generated by sdk to its function responsible for simulating swaps. It seems that not all of them were needed there.

Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts]: Removed unnecessary params

File: packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/BalancerFetcher.ts]: Removed

File: packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts (hunk 7 above)
Comment (cehali) [reply packages/oracle-node/src/fetchers/balancer/balancer-fetcher-configs.ts:45]: I tried to find something from where this difference come and couldn't find anything which would explain it

**Review by rpiszczatowski — DISMISSED** [review]

**Comment by koonopek** [comment]
@cehali slippage is merged, ping me after resolving conflicts I will review then @cehali

**Review by rpiszczatowski — APPROVED** [review]
Good job!
